Finalize - pending review - the terminolgy for the specification (#193)
hyperledger/anoncreds-spec: spec/terminology.md

## Terminology

[[def: Accumulator, accumulator]]

> ~ A [cryptographic accumulator] is used space- and time-efficient method of proving a value membership is in a set of values without revealing the individual members of the set. In AnonCreds v1, an accumulator is a core element of the [[ref: verifiable credential]] revocation mechanism.

Reviewer: ```suggestion
~ A [cryptographic accumulator] is used as a space- and time-efficient method of proving a value membership in a set of values without revealing the individual members of the set. In AnonCreds v1, an accumulator is a core element of the [[ref: verifiable credential]] revocation mechanism.
```

Reviewer: I would be tempted to reference the type of accumulator used, but maybe this is the wrong place

Author: What is the type of accumulator used? Do you know off-hand?  I can check if needed.

Author: Updated the text, and updated where the accumulator is used in AnonCreds — revocation.

Reviewer: It is an RSA based accumulator I believe, but I do not know the exact name offhand

[Hyperledger Indy]: https://www.hyperledger.org/projects/hyperledger-indy
[DID Core specification]: https://www.w3.org/TR/did-core/
[AnonCreds Methods Registry]: https://hyperledger.github.io/anoncreds-methods-registry/

[[def: AnonCreds Objects, AnonCreds objects]]

~ The published and shared data objects used in AnonCreds v1.0 including the published objects [[ref: schema]], [[ref: credential definition]], [[ref: revocation registry definition]], [[ref: revocation registry entry]] and the shared objects [[ref: credential offer]], [[ref: credential request]], and [[ref: presentation request]].

[[def: BigNumber, BigNumbers]]

~ A [BigNumber] is an data object which safely allows mathematical operations on numbers of any magnitude. BigNumbers are commonly used in cryptography schemes, including those underlying AnonCreds v1.0.

[BigNumber]: https://docs.ethers.org/v5/api/utils/bignumber/

[[def: Blinded Secret, Blinded secret, blinded secret, unblinded secret, Unblinded secret]]

> ~ A cryptographic technique where a secret value (a number) is blinded before it is shared such that the sender can later prove knowledge ("unblind") of the secret value. The AnonCreds v1.0 [[ref: link secret]]

Reviewer: ```suggestion
~ A cryptographic technique where a secret value (a number) is blinded before it is shared such that the sender can later prove knowledge of the secret value. The AnonCreds v1.0 [[ref: link secret]]

This is not what I understand as ublinding. Unblinding is ther removal of the blinding factor from a signature over a blinded value. Such that you get a signature on the unblinded value.

I think just removing as I have suggested is okay.
```

Author: Sounds good — thanks. Done.

mechanism is based on the use of a blinded secret.

[[def: Blinded Secrets Correctness Proof]]

~ A [[ref: ZKP]]-based proof that can be verified to show that a [[ref: blinded secret]] was produced correctly from an [[ref: unblinded secret]] without exposing the secret.

[[def: Blinding Factor, blinding factor]]

> ~ A blinding factor is a random 3152 bit selected from the set of integers up to the order of the RSA group, `n`. It is generated by the [[ref:holder]] to blind the their [[ref:link secret]] during credential issuance. The blinding factor can then be removed from the signature the [[ref:issuer]] produces to retrieve a valid signature over the [[ref: unblinded link secret]]. A blinding factor is similarly generated for each non-disclosed attribute during credential presentation, such that a [[ref: holder]] can prove they know these values without revealing them to the [[ref: verifier]].

Reviewer: ```suggestion
~ A blinding factor is a random [[ref: BigNumber]] selected from the set of integers up to the order of the RSA group, `n`. It is generated by the [[ref:holder]] to blind their [[ref:link secret]] during credential issuance. Knowledge of the blinding factor can be used to create a [[ref: Blinded Secrets Correctness Proof]].The blinding factor can be removed from the signature the [[ref:issuer]] produces to retrieve a valid signature over the [[ref: unblinded link secret]]. A blinding factor and associated [[ref: Blinded Secrets Correctness Proof]] are similarly generated for each non-disclosed attribute during credential presentation, such that a [[ref: holder]] can prove they know these values without revealing them to the [[ref: verifier]].
```

Author: Nice. Done.

[[def: Call Home, Call home, call home]]

> ~ Call home is a term used when evaluating the privacy characteristics of [[ref: verifiable credential]] deployments. If a [[ref: holder]] presenting data from a verifiable credential must always contact ("call home to") the [[ref: issuer]], [[ref: holder]] actions are open to the actual, or perception of, tracking of the holder by the issuer. [[ref: Verifiable credential]] schemes that do not make possible the tracking of [[ref: holder]] activities by [[ref: issuers]] are preferred.

Reviewer: Is it not also if a Verifier must call home to the issuer in order to verify each credential they are presented with?

Author: It is of far less concern, I think.  That a verifier uses a particular type of ID becomes known, but nothing about the holders they are verifying.  Given the existing interaction between the issuer/holder, that is what enables tracking. A verifier would not be anything but an arbitrary verifier to the issuer.  Its number of verifications would seem to be the most that could be tracked — with lots of techniques to get around that.

~ A claim is a part of digital identity related to a [[ref: subject]]. A claim can be attested by the identity subject itself, or it can be asserted by another entity.

[[def: Correlatability, correlatability, correlation, Correlation, correlatable, Correlatable]]

~ When a verifiable credential scheme that has the attribute of [[ref: unlinkability]], the data from the process of sharing a [[ref: verifiable presentations]] with different verifiers cannot be correlated to identify the [[ref: holder]].

[[def: Credential, Credentials, credential, credentials, verifiable credential, verifiable credentials, Verifiable Credential, Verifiable Credentials]]

~ A credential is a set of [[ref: claims]] about an identity [[ref: subject]]. A verifiable credential is a tamper-proof credential whose authorship is cryptographically verifiable. An anonymous credential, also known as AnonCreds, is a verifiable credential that has privacy-preserving properties to enable data minimization and correlation resistance.

[[def: Credential Definition, Credential Definitions, Public Credential Definition, Private Credential Definition, CredDef, CredDefs, credential definition, CLAIM_DEF, CRED_DEF]]

> ~ A credential definition (also known as CRED_DEF or CLAIM_DEF) contains data required for [[ref: credential]] issuance (used by the [[ref:issuer]]) as well as [[ref: credential]] validation data (used by the [[ref:holder]] and the [[ref:verifier]]). A credential definition is generated by the [[ref:issuer]] before [[ref:credential]] issuance and consists of two distinct but strongly related parts, the **Public Credential Definition** that is published for everyone to use and **Private Credential Definition** that is held as a secret by the issuer and has the private keys necessary to generate signed [[ref: verifiable credentials]]. A credential definition may be generated such that credentials it produces can be revoked.

Reviewer: Interesting. I always thought that CRED_DEF refers to only the public part. Here seems to be saying it is both the public and private parts.

Author: The “private” data of a Cred Def is just the corresponding data to the public parts, so doesn’t seem to be worth splitting up.

Author: Reworded.

[[def: Credential Key Correctness Proof]]

> ~ A proof generated during the creation of the [[ref: credential definition]] and included in the [[ref: credential offer]] so that the [[ref:holder]] can verify that the published [[ref: credential definition]] used in the blinding process belongs to the [[ref:issuer]].

Reviewer: Yeah see, this makes me thing that credential definition should be public. Then there is some other defined term for the private part.

> ~ A proof generated during the creation of the [[ref: credential definition]] and included in the [[ref: credential offer]] so that the [[ref:holder]] can verify that the published [[ref: credential definition]] used in the blinding process belongs to the [[ref:issuer]].

Author: OK - fair enough. I’ll try a reword...

> ~ A credential definition (also known as CRED_DEF or CLAIM_DEF) contains the public data required for [[ref: credential]] issuance (used by the [[ref:issuer]]) as well as [[ref: credential]] validation data (used by the [[ref:holder]] and the [[ref:verifier]]). A credential definition is generated by the [[ref:issuer]] before [[ref:credential]] issuance and published for anyone (primarily [[ref: holders]] and [[ref: verifiers]]) to use. In generating the public credential definition, related private data is also generated and held as a secret by the issuer. The secret data includes the private keys necessary to generate signed [[ref: verifiable credentials]] that can be presented and verified using the published credential definition. A credential definition can optionally be generated such that the credentials it produces can be revoked.

Reviewer 2: This should make clear that one cred def can be used for any number of issuances. It might be inferred that a cred def is required per issuance with the current reading.

Author: Updates made — good catch.

[[def: Credential Offer, credential offer, Credential offer]]

~ A credential offer is a data object sent by an [[ref: issuer]] to a [[ref: holder]] offering to issue a [[ref: credential]]. The credential offer contains the details about the claims the [[ref: issuer]] intends to issue to the [[ref: holder]]. A [[ref: holder]] can reply to the [[ref: issuer]] with a [[ref: credential request]]. A credential offer also includes a [[ref: nonce]] and a [[ref: Credential Key Correctness Proof]].

[[def: Credential Request, credential request, Credential request]]

> ~ A credential request is a request from an [[ref: holder]] towards a [[ref: issuer]] to get a credential issued by the [[ref: issuer]]. The credential request references a preceding [[ref: credential offer]] and defines the claims the [[ref: holder]] wants to get issued. A credential request also includes a [[ref: nonce]] that is later to use the issuance of the credential.

Reviewer: ```suggestion
~ A credential request is a request from an [[ref: holder]] towards a [[ref: issuer]] to get a credential issued by the [[ref: issuer]]. The credential request references a preceding [[ref: credential offer]] and defines the claims the [[ref: holder]] wants to get issued, including a [[ref: Blinded Secret]] and associated [[ref: Blinded Secrets Correctness Proof]]. A credential request also includes a [[ref: nonce]] that is later to use the issuance of the credential.
```

[[def: Data Minimization, data minimization, Data minimization]]

> ~ An attribute of verifiable data sharing schemes that considers privacy of a scheme based on the amount of data shared in a given interaction. Ideally, the minimum amount of data is shared for the business purpose of the interaction. Techniques such as [[ref: selective disclosure]], [[ref: predicates]], and [[ref: unlinkability]], all available in AnonCreds v1.0 support the goal of privacy-preserving, minimal data sharing.

Reviewer 2: several times the term `business purpose` is used in definitions. I'm not sure we should include the term `business`, and I think it reads fine without it.

Author: Removed.

[[def: DID, DIDs]]

> ~ A Decentralized Identifier (DID), defined by the [W3C DID Core Specification](https://w3c.github.io/did-core/), is a type of identifier that is useful in enabling verifiable, decentralized digital identity. A DID refers to any subject (e.g., a person, organization, thing, data model, abstract entity, etc.) as determined by the controller of the DID. DIDs are not used in AnonCreds itself but there must be a verifiable identifier (usually a DID) with an enforced relationship between [[ref: schema publishers]] and [[ref: issuers]] and the AnonCreds objects they publish. This is outlined in a note in [this specification section](#anoncreds-setup-data-flow).

Reviewer: In some earlier instances you defined the links serparetly. Do we want to continue that for the DID core link?

E.g. BigNumber is like this

[BigNumber]: https://docs.ethers.org/v5/api/utils/bignumber/

Author: Good catch — aligned to use the one link definition.

[[def: DID Method, DID method, DID Methods, DID methods]]

> ~ DID methods specify how DID documents are written (registered) and read (resolved) on a given [[ref: verifiable data registry]] implementation, allowing DIDs to be registered and resolved in a wide variety of storage containers. The capabilities required by DID Methods are defined in the [DID Core specification], and the (many) DID Methods are defined in the [DID Methods Registry] repository.

Reviewer: ```suggestion
~ DID methods specify how DID documents are created and read (resolved) on a given [[ref: verifiable data registry]] implementation, allowing DIDs to be registered and resolved in a wide variety of storage containers. The capabilities required by DID Methods are defined in the [DID Core specification], and the (many) DID Methods are defined in the [DID Methods Registry] repository.
```
```suggestion
~ DID methods specify how DID documents are written (registered) and read (resolved) on a given [[ref: verifiable data registry]] implementation, allowing DIDs to be created and resolved in a wide variety of storage containers. The capabilities required by DID Methods are defined in the [DID Core specification], and the (many) DID Methods are defined in the [DID Methods Registry] repository.
```

A DID document does not always have to be registered to be valid. E.g. did:key. I think created aligns better with DID spec

Author: Updated
